File: mdbook/src/mapreduce/fortran/openacc/optimized.f90
PROGRAM mapreduceoptimized
    integer, parameter :: num = 100000000
    integer*8 :: res = 0
    integer, dimension(:), allocatable :: elements
    real :: start_time, end_time

    allocate(elements(num))

    call cpu_time(start_time)
    !$acc data create(elements)
    !$acc parallel loop
    do i=1,num
        elements(i) = i
    enddo

    !$acc parallel loop reduction(+:res)
    do i=1,num
        res = res + elements(i)
    enddo
    !$acc end data
    call cpu_time(end_time)

    print *, "time:", end_time - start_time, "seconds"
    print *, "the result is:", res
END PROGRAM mapreduceoptimized

File: mdbook/src/mapreduce/cpp/openacc/optimized.cpp
#include <iostream>
#include <stdlib.h>
#include <timer.h>

using namespace std;

int main() {
  int num = 100000000;
  int *elements = new int[num];
  long long res = 0;

  timer time;

  /* ANCHOR: mapreduce */
  #pragma acc data create(elements[:num])
  {
    #pragma acc parallel loop
    for (int i = 0; i < num; i++) {
      elements[i] = i;
    }
    #pragma acc parallel loop reduction(+ : res)
    for (int i = 0; i < num; i++) {
      res += elements[i];
    }
  }
  /* ANCHOR_END: mapreduce */

  cout << "Elapsed time: " << time.getTime() << endl;

  cout << "The result is: " << res << endl;
}
